### ISSUE TITLE
Backward Pass of Gradient Penalty is missing in WGAN-GP

### ISSUE THREAD

**Issue by hello-fri-end** [mlpack/mlpack#3063]
File : https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/gan/wgangp_impl.hpp#L79

I'm reopening this issue since the [Dual-Optimizer PR](https://github.com/mlpack/mlpack/pull/2876) was closed. So, if someone wants, they can reopen that PR or work on this issue separately.

#### Issue description

WGAN-GP Loss function of Critic :
L = D(x) - D(G(z)) + lamda * ( norm(gradient(x')) -1) ^2

The backward pass of gradient penalty term is missing in the current implementation. 

 Following links may be helpful:
1. [numpy implementation of gradient penalty](https://github.com/ddbourgin/numpy-ml/blob/master/numpy_ml/neural_nets/losses/losses.py#L457)
2. The conversation in [Dual-Optimizer PR](https://github.com/mlpack/mlpack/pull/2876).
3. My [stack overflow post](https://stackoverflow.com/questions/69357832/deriving-equations-for-back-propagation-of-gradient-penalty).

**Comment by zoq** [comment]
Super, let us know if you need any further information.

**Comment by jeff10joy** [comment]
Hey I just recently started to work on this, have a couple of queries
1). I see in dual-optimizer PR, it is mentioned calculate gradient w.r.t mixed image batches, what exactly does this mean?
2). The wgan_imp file already has wgradient penalty implemented, am I missing something here? If yes, could you please let me know?
Thanking you.

**Comment by hello-fri-end** [comment]
> The wgan_imp file already has wgradient penalty implemented, am I missing something here? If yes, could you please let me know? Thanking you.

Hey Jeff! Thanks for working on this issue. Currently, only the forward pass of gradient penalty has been implemented while backward pass is missing. 

> I see in dual-optimizer PR, it is mentioned calculate gradient w.r.t mixed image batches, what exactly does this mean?


The Gradient() method of the FFN class calculates gradient with respect to parameters of neural network. However, the gradient term in gradient penalty refers to gradient with respect to the images in the mixed batch where mixed batch (let's call it x')   is given by:

`x' = ε * real_batch + ( 1 - ε ) * generated_batch`

Here's the pytorch code for Gradient Penalty:

```
       epsilon_shape = [real_data.shape[0]] + [1]*(real_data.dim() - 1)
        epsilon = torch.rand(epsilon_shape)
        epsilon = epsilon.to(fake_data.device, fake_data.dtype)
        real_data = real_data.to(fake_data.dtype)
        x_hat = epsilon * real_data + (1-epsilon) * fake_data.detach()
        x_hat.requires_grad = True
        logits = self.discriminator(x_hat, condition, landmarks)
        logits = logits.sum()
        grad = torch.autograd.grad(
            outputs=logits,
            inputs=x_hat,
            grad_outputs=torch.ones(logits.shape).to(fake_data.dtype).to(fake_data.device),
            create_graph=True
        )[0]


```
Notice that the input passed to autograd is  ` x_hat`. Let me know if I should clarify anything further. :)

**Comment by Sirish07** [comment]
I would like to take up this issue.

**Comment by shubham1206agra** [comment]
@hello-fri-end Please add the task of removing the boost::visitor and replace with vtable #2777 

So that future work can be done now itself.

**Comment by Sirish07** [comment]
Hey @hello-fri-end ,

From what I understand, we just need to implement this [part](https://github.com/ddbourgin/numpy-ml/blob/1e10697258952865e497a797e16558c40087d6fc/numpy_ml/neural_nets/losses/losses.py#L497) right?

**Comment by hello-fri-end** [comment]
> Hey @hello-fri-end ,
> 
> From what I understand, we just need to implement this [part](https://github.com/ddbourgin/numpy-ml/blob/1e10697258952865e497a797e16558c40087d6fc/numpy_ml/neural_nets/losses/losses.py#L497:L503) right?

Yes, you're right.
